### PR TITLE
Add direction classification algorithm for all connections

### PR DESCRIPTION
**Pull request by stb2132**
### What does this PR do?

This PR adds an algorithm to determine connection direction by utilizing the current state of all connections. Since we use direction to determine if we should skip a connection, it also refactors code for skipping connection out of the getActiveConnections() and getInitPerfPolling() functions into the getConnections() function. 

### Motivation

We currently cannot properly classify connections as local since we do not compare a connection to the others, so these changes allow us to do this classification. 

### Additional Notes

This change will change some DNS connections from being classified as `NONE` to `LOCAL`. For  example, previously a connection with addresses like this:

```
 {
      "pid": 25991,
      "laddr": {
        "ip": "127.0.0.1",
        "port": 47995,
        "containerId": "",
        "hostId": 0
      },
      "raddr": {
        "ip": "127.0.0.1",
        "port": 8125,
        "containerId": "",
        "hostId": 0
      },
      "type": "udp",
      ...
    }
  ],
  "dns": {}
```

would have direction `NONE`, but will now be specified as `LOCAL`.

This change fixes the unit tests:

- TestTCPSendAndReceive
- TestPreexistingConnectionDirection 
- TestTCPOverIPv6
- TestLocalDNSCollectionDisabled 
- TestShouldSkipBlacklistedConnection

### CODE REVIEW

File: pkg/ebpf/tracer.go
Comment (leeavital) [inline pkg/ebpf/tracer.go]: let's call this `removeBlacklistedConnections` since the connections being removed here are not expired.

File: pkg/ebpf/tracer.go
Comment (leeavital) [inline pkg/ebpf/tracer.go]: This should be `conn.IPTranslation.ReplSrcIP` because the fields for IPTranslation are in reverse

File: pkg/ebpf/tracer.go
Comment (leeavital) [inline pkg/ebpf/tracer.go]: I don't think this part is necessary, everything in `hit` will end up getting processed once as part of the outer loop.   That means that the `connRAddrMap` can be a map from `util.Address` to `struct{}` which will save on space.

File: pkg/ebpf/tracer.go
Comment (kevinconaway) [inline pkg/ebpf/tracer.go]: Should this be the `ReplSrcIP`? (cc @leeavital)

File: pkg/ebpf/tracer.go
Comment (kevinconaway) [inline pkg/ebpf/tracer.go]: nit, if this is a collection, would you please name it `hits` or perhaps something like `localRaddrs`?

Comment (leeavital) [inline pkg/ebpf/tracer.go]: This part should also be IPtranslation aware, and use IPTranslation.ReplDstIP if available

File: pkg/ebpf/tracer.go
Comment (kevinconaway) [inline pkg/ebpf/tracer.go]: nit, it might simplify things to alias `connections[i]` to a local var in this loop to prevent mistakes since you have an inner loop

**Review by kevinconaway — COMMENTED** [review]
It doesn't look like `go fmt` was run on these files.  Would you please do so?

I'm surprised the agent builds without it.

File: pkg/ebpf/tracer.go
Comment (stb2132) [reply pkg/ebpf/tracer.go]: Thanks for the note - going to be removing the inner loop, but will keep this in mind moving forward.

File: pkg/ebpf/tracer.go
Comment (sunhay) [inline pkg/ebpf/tracer.go]: Why? Do we know how many more connections we'll be storing in state? I've seen 10k+ local dns connections before, which I'd prefer to leave out of the state...

File: pkg/ebpf/tracer.go
Comment (sunhay) [inline pkg/ebpf/tracer.go]: `RAddr` -> `useRAddrAsKey` so you don't need the comment below

File: pkg/ebpf/tracer.go
Comment (sunhay) [inline pkg/ebpf/tracer.go]: `key := connKey{ Type: connStat.Type}`

File: pkg/ebpf/tracer.go
Comment (sunhay) [reply pkg/ebpf/tracer.go]: I wasn't aware that we were using direction to skip connections, when was that added?

File: pkg/ebpf/tracer.go
Comment (sunhay) [inline pkg/ebpf/tracer.go]: I thought we were going to get rid of `Direction = LOCAL` as a concept?

File: pkg/ebpf/tracer.go
Comment (sunhay) [inline pkg/ebpf/tracer.go]: nit: this should just be an `else if` on the above block

File: pkg/ebpf/tracer.go
Comment (sunhay) [inline pkg/ebpf/tracer.go]: `for _, i := range connections`?

File: pkg/ebpf/tracer.go
Comment (sunhay) [inline pkg/ebpf/tracer.go]: `for _, i := range connections`

File: pkg/ebpf/tracer.go
Comment (leeavital) [reply pkg/ebpf/tracer.go]: This was moved because we now need the full list of connections to determine connection direction and we only have one available here.

shouldSkipConnection is defined as

```
func (t *Tracer) shouldSkipConnection(conn *ConnectionStats) bool {
	isDNSConnection := conn.DPort == 53 || conn.SPort == 53
	if !t.config.CollectLocalDNS && isDNSConnection && conn.Direction == LOCAL {
		return true
	} else if IsBlacklistedConnection(t.sourceExcludes, t.destExcludes, conn) {
		return true
	}
	return false
}
```

The only place it uses Local is to enfore the older CollectLocalDNS option which was replaced by connection blacklisting.  How about removing the `conn.Direction == LOCAL` check and going back to calling `shouldSkipConnection` before a connection gets stored in `t.state`

File: pkg/ebpf/tracer.go
Comment (leeavital) [reply pkg/ebpf/tracer.go]: Or even:

```
for _, conn := range connections {
   connLadddrMap[newConnKey(conn, false)] = struct{}{}
}
```

File: pkg/ebpf/tracer.go
Comment (leeavital) [inline pkg/ebpf/tracer.go]: I don't think this variable needs to be declared outside the for loop immediately below this line.

File: pkg/ebpf/tracer.go
Comment (leeavital) [inline pkg/ebpf/tracer.go]: we can call this `connLAddrs`  or even `laddrs`. No need to put `map` in the name since that's part of the type

File: pkg/ebpf/tracer_test.go
Comment (leeavital) [inline pkg/ebpf/tracer_test.go]: nit: put this import with the rest of the "github.com/DataDog/..." imporrts

File: pkg/ebpf/tracer_test.go
Comment (leeavital) [inline pkg/ebpf/tracer_test.go]: nit: these are more clearly written as a literal

```
connStats := ConnectionStats{
   Source: util.AddressFromString("10.0.2.25"),
   Sport: 5223,
   Dest: ...
}
```

File: pkg/ebpf/tracer_test.go
Comment (leeavital) [inline pkg/ebpf/tracer_test.go]: nit: this is better written as `require.NoErr(t, err)

File: pkg/ebpf/tracer_test.go
Comment (leeavital) [inline pkg/ebpf/tracer_test.go]: nit: instead of using `NewTracer()`  which will attach a BPF program, you can do something like:

```go
tr := &tracer{
   // initialize anything necessary here
}
```

And move this to a test file that isn't tagged as `linux_bpf`. That way these tests will run as part of a normal `go test`

File: pkg/ebpf/tracer.go
Comment (leeavital) [inline pkg/ebpf/tracer.go]: nit: I suggest writing this as an `if/else if/else` and avoiding `continue`

File: pkg/ebpf/tracer_test.go
Comment (leeavital) [inline pkg/ebpf/tracer_test.go]: For these four similar tests, you may be able to use a table test to avoid copying the same setup across several tests.

File: pkg/ebpf/tracer.go
Comment (sunhay) [inline pkg/ebpf/tracer.go]: Hmm we'll want to make sure we add an additional check to confirm this is a local DNS connection if we're removing `conn.Direction == LOCAL`

Comment (sunhay) [reply pkg/ebpf/tracer.go]: Possibly looking at the IP to see if it's in the LAN subnets

Comment (leeavital) [reply pkg/ebpf/tracer.go]: https://github.com/DataDog/datadog-agent/pull/4727/files will fix this.

File: pkg/ebpf/tracer_test.go
Comment (stb2132) [reply pkg/ebpf/tracer_test.go]: Moved other tests to a file and do note enable a full tracer. This test remains as an integration test.

File: pkg/ebpf/tracer.go
Comment (leeavital) [inline pkg/ebpf/tracer.go]: You can leave out the 0 here.

**Review by leeavital — COMMENTED** [review]
LGTM pending us referencing a tag of agent-payload, and not a hash

File: pkg/ebpf/tracer.go
Comment (stb2132) [reply pkg/ebpf/tracer.go]: Removed this - looks like it doesn't update here though. Resolving conversation.

File: pkg/ebpf/tracer.go
Comment (kevinconaway) [inline pkg/ebpf/tracer.go]: nit, does this need to be done in the tracer or can it be done in `Connections` ?

File: pkg/ebpf/tracer.go
Comment (kevinconaway) [inline pkg/ebpf/tracer.go]: We don't need to take the IPTranslation in to account for the source IP?

Comment (leeavital) [reply pkg/ebpf/tracer.go]: No, because of how NAT connections show up for intra-host connections, the original IP is the "real" IP and the NATed source is the "virtual" IP.

File: pkg/ebpf/tracer.go
Comment (leeavital) [reply pkg/ebpf/tracer.go]: Does this actually fit better in `state.Connections` ? `tracer` is already doing a bunch of post-processing (like  handling connection direction, ip translations, and DNS). This seems to fit in nicely.

Comment (kevinconaway) [reply pkg/ebpf/tracer.go]: 🤷‍♂ my thought was that since the method does nothing specific with the tracer data (the method itself is stateless) it would be better off elsewhere so that other calls to `Connections` would have the data.  

I don't have strong feelings though.

Comment (stb2132) [reply pkg/ebpf/tracer.go]: I've been playing around with this, and it can be done in `Connections` fairly simply. I have the changes on my local branch and will push them up.

**Review by kevinconaway — DISMISSED** [review]
Before merging, please make a release of the agent-payload and update the PR with the actual version.

**Comment by stb2132** [comment]
Merged change in agent-payload:

- https://github.com/DataDog/agent-payload/releases/tag/4.24.0

**Review by leeavital — APPROVED** [review]

**Review by derekwbrown — APPROVED** [review]
should I ask what happened to 20-23?
LGTM

**Pull request merged**
